src/odemis/acq/acqmng.py

```diff
@@ -83,6 +84,189 @@ def acquire(streams, settings_obs=None):
     return future
 
 
+def acquireZStack(streams, zlevels, settings_obs=None):
+    """
+    The acquisition manager of a zstack
+    streams (list of Stream): the streams to be acquired
+    zlevels (dict Stream -> list of floats): a dictionary containing the streams and the
+    corresponding lists of actuator z positions
+    settings_obs (SettingsObserver):
+    return (ProgressiveFuture): the future that will be executing the task
+    """
+    # TODO settings observer
+    # create future
+    future = model.ProgressiveFuture()
+    # create acquisition task
+    acqui_task = ZStackAcquisitionTask(future, streams, zlevels, settings_obs)
+    # add the ability of cancelling the future during execution
+    future.task_canceller = acqui_task.cancel
+
+    # set the progress of the future
+    total_duration = acqui_task.estimate_total_duration()
+    future.set_end_time(time.time() + total_duration)
+
+    # assign the acquisition task to the future
+    executeAsyncTask(future, acqui_task.run)
+
+    return future
+
+
+class ZStackAcquisitionTask(object):
+    """
+    This class represents an acquisition task for a zstack of data.
+    """
+
+    def __init__(self, future, streams, zlevels, settings_obs):
+        """
+        The class constructor
+        future (ProgressiveFuture): the future that will execute the task asynchronously
+        streams (list of Stream): the streams to be acquired
+        zlevels (dict Stream -> list of floats): a dictionary containing the streams and the
+        corresponding lists of actuator z positions
+        settings_obs (SettingsObserver):
+        """
+        self._main_future = future
+        self._future_state = None
+        self._streams = sortStreams(streams)
+        self._zpos = next((l for l in zlevels.values() if len(l) > 1), None)
+        self._zlevels = zlevels
+        self._settings_obs = settings_obs
+        self._zstep_duration = self._estimate_zstep_duration()
+
+    def cancel(self, _):
+        """
+        Called when the user presses the button "cancel" and the future is still running.
+        (but there might  be some delay (not instantly called) )
+        return (Boolean): True if canceled, False otherwise
+        """
+        self._future_state = CANCELLED
+        # cancel the sub-futures
+        self._actuator_f.cancel()
+        self._single_acqui_f.cancel()
+        return True
```

Comment:
I think this part usually also allows to return a False if the future already finished. Like in "autofocus.py" - _cancelSparc2ManualFocus line 767.
But maybe that is a different part of the future (I didn't have time to check that extensively)

Reply:
Indeed, you can also check `_cancelAcquisition` of the `TiledAcquisitionTask`.

Reply:
I had a look at the implementations that you, both, mentioned. Also in the implementation of the future class in the concurrent library they do this. But why actually? I don't totally get why we return False or True? Since we do not use the return value in the acquisition controller. Do you have explanation maybe?

Reply:
The return value is for confirming cancellation. If the movement is finished, then cancel didn't do anything and it shouldn't say it's  cancelled (ie True), right?